Consul on ECS 0.4.0

cthain opened the pull request:
Changes proposed:
* Add documentation for Admin Partitions and Namespaces
* Fix some minor typos

pglass commented on website/content/docs/ecs/enterprise.mdx:
```diff
@@ -59,7 +59,72 @@ features.
+the partition it will default to the `default` admin partition. Similarly, if a `mesh-task` does
+not define the namespace it will default to the `default` namespace.
+
+To support admin partitions and namespaces, a `mesh-task` must use a Consul Enterprise image.
+By default the `mesh-task` module uses a Consul OSS image. The module provides
+a `consul_image` field that can be used to set the Consul client image to an Enterprise version.
```
```suggestion

```

The enterprise image requirement is mentioned at the top of the page, so I'm thinking we don't need to restate it here. Thoughts?

pglass commented on website/content/docs/ecs/terraform/secure-configuration.mdx:
```diff
@@ -21,7 +21,7 @@ A secure Consul cluster should include the following:
 for tasks on the service mesh. To learn more about the ACL Controller, please see [Automatic ACL Token Provisioning](/docs/ecs/architecture#automatic-acl-token-provisioning).
 
-To deploy the controller, you will first need store an ACL token with `acl:write` privileges
+To deploy the controller, you will first need to store an ACL token with `acl:write` privileges
```
Incidentally, we also need the controller's token to have `operator:write` now that it creates partitions and namespaces

cthain replied:
Good catch!

pglass commented:
Also, there's a [Task Tags](https://www.consul.io/docs/ecs/manual/install#task-tags) section that we should update with the new partition and namespace tags.

cthain commented on website/content/docs/ecs/architecture.mdx:
```diff
@@ -95,3 +95,20 @@ If the following conditions apply, ECS health checks automatically sync with Con
+## Admin Partitions and Namespaces
+
+Consul on ECS supports admin partitions and namespaces.
+When [Admin Partitions and Namespaces](/docs/ecs/enterprise#admin-partitions-and-namespaces) are enabled,
```
:+1: agreed. This comment compliments https://github.com/hashicorp/consul/pull/12694/files#r842860413.

pglass commented on website/content/docs/ecs/architecture.mdx:
```diff
@@ -95,3 +95,20 @@ If the following conditions apply, ECS health checks automatically sync with Con
 every container that fits these criteria and the `health-sync` container ensures
 that the ECS and Consul health checks remain in sync.
+
+## Admin Partitions and Namespaces
```
Yeah, I think combining and renaming the heading is good.

(Remember to update incoming links if you remove/change headings)

trujillo-adam commented on website/content/docs/ecs/requirements.mdx:
```diff
@@ -13,6 +13,11 @@ The following requirements must be met in order to install Consul on ECS:
 1. **Subnets:** ECS Tasks can run in private or public subnets. Tasks must have [network access](https://aws.amazon.com/premiumsupport/knowledge-center/ecs-pull-container-api-error-ecr/) to Amazon ECR or other public container registries to pull images.
```
Are these requirements intended to be performed sequentially? Is there a reason that they are in an ordered list? It might be a little less overwhelming for users if we grouped like requirements into sections, e.g., "ACL Controller Requirements" and "Sidecar Requirements". Thoughts?

cthain replied:
Good question. I'm not certain what the original intention was, I just followed the ordered list because that's what was already there. I don't think that the items in this list need to be satisfied in any particular order. I'm fine with separate sections, although they would be quite sparse in most cases. Tagging @pglass and @erichaberkorn in case they have opinions/suggestions/thoughts.

pglass replied:
Yeah, this could be an unordered list. They aren't ordered steps, or even a prioritized list.

cthain replied:
I have changed this to an unordered list.

cthain commented on website/content/docs/ecs/architecture.mdx:
```diff
@@ -84,6 +90,21 @@ token does not yet exist.
+tokens for tasks in a single admin partition and supports one ACL controller instance per ECS
+cluster. This results in an architecture with one admin partition per ECS cluster.
+
+The ACL controller automatically performs the following:
+* Create its configured admin partition at startup if it does not exist.
+* Inspect ECS task tags to discover a task's intended partition and namespace.
+  The ACL controller ignores tasks which do not have a matching partition tag.
+* Create namespaces when tasks startup. Namespaces are only created if the do not exist.
+* Provision ACL tokens and ACL policies that are scoped to the applicable admin partition and namespace.
+* Provision ACL tokens that allow services to communicate with upstreams across admin partitions and namespaces.
```
Yes that's correct. The configuration is handled by the ACL controller Terraform module.

cthain commented on website/content/docs/ecs/enterprise.mdx:
```diff
@@ -38,16 +38,16 @@ run Consul Enterprise clients then you must enable ACLs.
-Consul supports running Consul Enterprise servers with Consul OSS (Open Source) clients. Since
-currently no Consul Enterprise features are supported that require Consul client support,
-you can run Consul OSS clients with Consul Enterprise servers without issue.
+Consul supports running Consul Enterprise servers with Consul OSS (Open Source) clients.
+Consul OSS clients can run with Consul Enterprise servers without issue as long as you are
+not using features that require Consul Enterprise client support.
+Admin Partitions and Namespaces require Consul Enterprise clients.
```
Yes thank you!

cthain commented on website/content/docs/ecs/architecture.mdx:
```diff
@@ -95,3 +95,20 @@ If the following conditions apply, ECS health checks automatically sync with Con
+
+When the ACL Controller provisions tokens for the clients and services, it ensures that those tokens are scoped to the applicable admin partition and namespace.
+The ACL Controller creates service tokens with attached policies that allow services to communicate with upstreams
+in different namespaces and admin partitions.
```
This list is not really attempting to describe an ordered set of operations so I think I would prefer to leave it unordered.

pglass commented on website/content/docs/ecs/enterprise.mdx:
```diff
@@ -59,7 +56,68 @@ features.
 | Network Segments                  | No            | Currently there is no capability to configure the network segment Consul clients on ECS run in.                                                                         |
-| Namespaces                        | No            | Currently there is no capability to configure the Consul namespace for a service on ECS.                                                                                |
-| Admin Partitions                  | No*           | Supported if Consul ECS clients run in the default partition. Otherwise there is currently no capability to configure the admin partition Consul clients in ECS run in. |
+| Namespaces                        | Yes           | This feature requires Consul Enterprise servers and clients.                                                                                                            |
```
nit: One exception is OSS clients can register into the default namespace and partition. So it might be good to add that caveat, maybe just as a note somewhere?

cthain replied:
Good call. I have added a note to each row.

cthain commented:
@trujillo-adam, @erichaberkorn, @pglass I've addressed all the review comments. Is this PR good to merge? Is the "Vercel - consul" pipeline failure something that needs to be sorted out or is that a known thing?

pglass commented:
I'm not sure on the Vercel error, but it seems unrelated to your changes.

Other PRs don't seem to have that issue @cthain do you maybe need to update/rebase this PR with main?

hc-github-team-consul-core commented:
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/630636.

hc-github-team-consul-core commented:
:cherries::white_check_mark: Cherry pick of commit 1502936c123e855f984f9589cf588153626b5ac0 onto `stable-website` succeeded!

hc-github-team-consul-core commented:
:cherries::white_check_mark: Cherry pick of commit 1502936c123e855f984f9589cf588153626b5ac0 onto `release/1.12.x` succeeded!
